### PR TITLE
Added build templates

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixed [VSTS-128](https://jira.sonarsource.com/browse/VSTS-128)

### CODE REVIEW

File: extensions/sonarcloud/templates/gradle/template.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
+{
+  "id": "sonarcloud_gradle_template",
```
Comment (duncanp-sonar) [inline extensions/sonarcloud/templates/gradle/template.json]: A build template is just a _.json_ file listing the tasks to be executed, and the property values to be pre-set for those tasks. The template file must be called _template.json_. There is no executable code.
I've added four templates: Maven, Gradle, VS Build and .NET Core build.

File: extensions/sonarcloud/templates/gradle/template.json [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        "timeoutInMinutes": 0,
+        "task": {
+          "id": "14d9cde6-c1da-4d55-aa01-2965cd301255",
+          "versionSpec": "1.*",
+          "definitionType": "task"
+        },
+        "inputs": {
+          "scannerMode": "Other",
+          "configMode": "file",
+          "configFile": "sonar-project.properties",
```
Comment (henryju) [inline extensions/sonarcloud/templates/gradle/template.json]: configMode and configFile are not visible when scannerMode = 'Other'. Do you really have to define a value in the template?

File: extensions/sonarcloud/templates/gradle/template.json [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+          "cliProjectKey": "",
+          "projectKey": "",
+          "cliProjectName": "",
+          "projectName": "",
+          "cliProjectVersion": "1.0",
+          "projectVersion": "1.0",
+          "cliSources": ".",
```
Comment (henryju) [inline extensions/sonarcloud/templates/gradle/template.json]: Same for all the previous props. I would find it more readable to not specify them in the template (if possible).

File: extensions/sonarcloud/templates/gradle/template.json [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+          "versionSpec": "2.*",
+          "definitionType": null
+        },
+        "inputs": {
+          "wrapperScript": "$(Parameters.wrapperScript)",
+          "tasks": "$(Parameters.tasks)",
+          "codeCoverageTool": "JaCoCo",
+          "sqAnalysisEnabled": "true",
+          "sqGradlePluginVersionChoice": "specify",
+          "sqGradlePluginVersion": "2.6.1"
```
Comment (henryju) [inline extensions/sonarcloud/templates/gradle/template.json]: 2.6.2 is the latest version

File: extensions/sonarcloud/templates/maven/template.json [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          "scannerMode": "Other",
+          "configMode": "file",
+          "configFile": "sonar-project.properties",
+          "cliProjectKey": "",
+          "projectKey": "",
+          "cliProjectName": "",
+          "projectName": "",
+          "cliProjectVersion": "1.0",
+          "projectVersion": "1.0",
+          "cliSources": ".",
```
Comment (henryju) [inline extensions/sonarcloud/templates/maven/template.json]: Same comment than for Gradle, don't specify hidden props if possible.

File: extensions/sonarcloud/templates/maven/template.json [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        "alwaysRun": false,
+        "timeoutInMinutes": 0,
+        "task": {
+          "id": "ac4ee482-65da-4485-a532-7b085873e532",
+          "versionSpec": "2.*",
+          "definitionType": null
+        },
+        "inputs": {
+          "mavenPOMFile": "$(Parameters.mavenPOMFile)",
+          "goals": "package",
```
Comment (henryju) [inline extensions/sonarcloud/templates/maven/template.json]: I wonder if it is good to specify values here. If you don't, will it fallback to use default value? If yes, then in will be better to not fix the value in the template, but instead rely on the task defaults. This way, if they later do some changes to their default, our template will stay "up to date".

File: extensions/sonarcloud/templates/gradle/template.json (hunk 4 above)
Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/gradle/template.json]: Done

File: extensions/sonarcloud/templates/gradle/template.json (hunk 3 above)
Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/gradle/template.json]: I've removed all of the properties except "scannerMode" from all of the templates.

File: extensions/sonarcloud/templates/maven/template.json (hunk 6 above)
Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/maven/template.json]: I've removed _goals_ but left _mavenPOMFile_ as it refers to a process level-variable.

File: extensions/sonarcloud/templates/gradle/template.json [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        "continueOnError": false,
+        "alwaysRun": false,
+        "displayName": "Publish Quality Gate Result",
+        "timeoutInMinutes": 0,
+        "task": {
+          "id": "38b27399-a642-40af-bb7d-9971f69712e8",
+          "versionSpec": "1.*",
+          "definitionType": "task"
+        },
+        "inputs": { "pollingTimeoutSec": "300" }
```
Comment (henryju) [inline extensions/sonarcloud/templates/gradle/template.json]: What about removing this input from the template?

File: extensions/sonarcloud/templates/maven/template.json [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        "continueOnError": false,
+        "alwaysRun": false,
+        "displayName": "Publish Quality Gate Result",
+        "timeoutInMinutes": 0,
+        "task": {
+          "id": "38b27399-a642-40af-bb7d-9971f69712e8",
+          "versionSpec": "1.*",
+          "definitionType": "task"
+        },
+        "inputs": { "pollingTimeoutSec": "300" }
```
Comment (henryju) [inline extensions/sonarcloud/templates/maven/template.json]: Same

File: extensions/sonarcloud/templates/netcore/template.json [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        "displayName": "Publish Quality Gate Result",
+        "timeoutInMinutes": 0,
+        "condition": "succeeded()",
+        "task": {
+          "id": "38b27399-a642-40af-bb7d-9971f69712e8",
+          "versionSpec": "1.*",
+          "definitionType": "task"
+        },
+        "inputs": {
+          "pollingTimeoutSec": "300"
```
Comment (henryju) [inline extensions/sonarcloud/templates/netcore/template.json]: Same

File: extensions/sonarcloud/templates/netcore/template.json [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+          "definitionType": "task"
+        },
+        "inputs": {}
+      },
+      {
+        "environment": {},
+        "enabled": false,
+        "continueOnError": false,
+        "alwaysRun": false,
+        "displayName": "Publish Quality Gate Result",
```
Comment (henryju) [inline extensions/sonarcloud/templates/netcore/template.json]: Same than for inputs, what if you don't specify a displayName?

File: extensions/sonarcloud/templates/netdesktop/template.json [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        "displayName": "Publish Quality Gate Result",
+        "timeoutInMinutes": 0,
+        "condition": "succeeded()",
+        "task": {
+          "id": "38b27399-a642-40af-bb7d-9971f69712e8",
+          "versionSpec": "1.*",
+          "definitionType": "task"
+        },
+        "inputs": {
+          "pollingTimeoutSec": "300"
```
Comment (henryju) [inline extensions/sonarcloud/templates/netdesktop/template.json]: same

File: extensions/sonarcloud/templates/gradle/template.json (hunk 7 above)
Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/gradle/template.json]: Done

File: extensions/sonarcloud/templates/netcore/template.json (hunk 10 above)
Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/netcore/template.json]: It seems to use the _friendlyName_ from the _task.json_. I've removed the _displayName_ for all of our tasks.
I've left the display names for the Microsoft tasks in cases where the corresponding default VSTS template has a display name (i.e. for the NuGet restore step and _dotnet_ tasks).

Comment (henryju) [reply extensions/sonarcloud/templates/netcore/template.json]: I would also remove for Microsoft tasks except if we see a benefit to not show the default friendly name. But that's minor.

Comment (duncanp-sonar) [reply extensions/sonarcloud/templates/netcore/template.json]: All of our templates are closely based on one of the default existing VSTS templates. I deliberately didn't change the definitions of the tasks copied from the standard templates to make it easier to compare our templates against the default templates in the future.

As it happens, the display name is required in three out of the four Microsoft tasks (the same task is run multiple times but with different parameters).

**Pull request merged**